src/index/storage.rs

```diff
@@ -66,3 +85,99 @@ impl FileSystemStorage {
         Ok(())
     }
 }
+
+#[derive(Debug)]
+pub struct S3Storage {
+    runtime: Runtime,
+    client: S3Client,
+    bucket: String,
+    key: String,
+}
+
+impl S3Storage {
+    fn new(bucket: &str, key: &str) -> Result<Self> {
+        let runtime = Runtime::new()?;
+
+        let config = runtime.block_on(async {
+            let global_config = aws_config::load_from_env().await;
+            let global_s3 = S3Client::new(&global_config);
+
+            let location = global_s3
+                .get_bucket_location()
+                .bucket(bucket)
+                .send()
+                .await?;
+            let region = location
+                .location_constraint()
+                .map(|c| c.as_str())
+                .unwrap_or("us-east-1")
+                .to_string();
+
+            info!("using S3 bucket {bucket} in region {region}");
+
+            let regional_config = aws_config::from_env()
+                .region(Region::new(region))
+                .load()
+                .await;
+
+            Ok::<_, anyhow::Error>(regional_config)
+        })?;
+        let client = S3Client::new(&config);
+
+        Ok(S3Storage {
+            runtime,
+            client,
+            bucket: bucket.into(),
+            key: key.into(),
+        })
+    }
+
+    fn read(&self) -> Result<Option<Index>> {
+        self.runtime.block_on(async {
+            let result = self
+                .client
+                .get_object()
+                .bucket(&self.bucket)
+                .key(&self.key)
+                .send()
+                .await;
+
+            match result {
+                Ok(response) => {
+                    // FIXME: this buffers the downloaded data into memory before deserializing it,
+                    // as I'm not aware of a way to convert from AsyncRead to Read.
```

Comment:
You can't convert AsyncRead to Read easily, but we can avoid copying it into a Vec manually -- `response.body.collect().await.reader()` will give you a `impl Read`. This is also slightly more efficient for large files as we're avoiding a single contiguous buffer, rather using the underlying segments that we get over the network.

But this is also fine, up to you if you want to change this.

Reply:
Changed.